README.md
# jzqt

docs/releases.rst
Release notes
=============

.. toctree::
    :maxdepth: 2

    releases/v0.2.0
    releases/v0.1.0


tests/test_timetools.py
# coding: utf-8

from datetime import datetime, date, timedelta

import pytest

import jzqt.timetools as jt


class TestTimerange(object):

    timerange_samples = [
        jt.timerange(date(1, 1, 1), date(1, 1, 1), jt.DAY),
        jt.timerange(date(1, 1, 1), date(1, 1, 1), -jt.DAY),
        jt.timerange(date(1, 1, 1), date(1, 1, 2), jt.DAY),
        jt.timerange(date(1, 1, 2), date(1, 1, 1), -jt.DAY),
        jt.timerange(date(1, 1, 1), date(1, 1, 1), jt.DAY),
        jt.timerange(date(2, 2, 2), date(1, 1, 1), jt.DAY),
        jt.timerange(date(1, 1, 1), date(2, 2, 2), -jt.DAY),
        jt.timerange(datetime(1, 1, 1), datetime(1, 1, 1, 1), jt.DAY),
        jt.timerange(datetime(1, 1, 2), datetime(1, 1, 1), -jt.DAY),
        jt.timerange(datetime(1, 1, 1, 1), datetime(1, 1, 1, 1), jt.DAY),
        jt.timerange(datetime(1, 1, 1), datetime(2, 2, 2), -jt.DAY),
        jt.timerange(datetime(2, 2, 2), datetime(1, 1, 1), jt.DAY),
    ]

    def test_init_raise_value_error(self):
        with pytest.raises(ValueError) as exc_info:
            jt.timerange(datetime(2017, 1, 1), datetime(2017, 2, 2), timedelta())
        msg = 'timerange() arg 3 must not be timedelta(0)'
        assert exc_info.value.args == (msg,)

    @pytest.mark.parametrize('func, arg_pos', [
        (lambda: jt.timerange(100, datetime(2017, 2, 3)), 1),
        (lambda: jt.timerange(date(2017, 2, 3), 4), 2),
        (lambda: jt.timerange(date(2017, 2, 3), date(2017, 4, 5), 86400), 3),
    ])
    def test_init_raise_type_error(self, func, arg_pos):
        with pytest.raises(TypeError) as exc_info:
            func()
        arg_type = 'timedelta' if arg_pos == 3 else 'date or datetime'
        assert exc_info.value.args == (
            'timerange() arg {} must be {} object'.format(arg_pos, arg_type),
        )

    @pytest.mark.parametrize('start, stop, step, results', [
        (date(1970, 1, 1), date(1970, 2, 1), 15 * jt.DAY,
         [date(1970, 1, 1), date(1970, 1, 16), date(1970, 1, 31)]),

        (date(1, 1, 1), date(2, 2, 2), 365 * jt.DAY,
         [date(1, 1, 1), date(2, 1, 1)]),

        (date(2, 2, 2), date(1, 1, 1), -365 * jt.DAY,
         [date(2, 2, 2), date(1, 2, 2)]),

        (date(1, 1, 1), date(1, 1, 1), jt.DAY, []),
        (date(1, 1, 1), date(1, 1, 1), -jt.DAY, []),
    ])
    def test_iter_date(self, start, stop, step, results):
        assert list(jt.timerange(start, stop, step)) == results

    @pytest.mark.parametrize('start, stop, step, results', [
        (datetime(1970, 1, 1), datetime(1970, 1, 1, 23), 10 * jt.HOUR,
         [datetime(1970, 1, 1), datetime(1970, 1, 1, 10),
          datetime(1970, 1, 1, 20)]),

        (datetime(1, 1, 1, 1), datetime(2, 2, 2, 2), 365 * jt.DAY,
         [datetime(1, 1, 1, 1), datetime(2, 1, 1, 1)]),

        (datetime(2, 2, 2, 2), datetime(1, 1, 1, 1), -365 * jt.DAY,
         [datetime(2, 2, 2, 2), datetime(1, 2, 2, 2)]),

        (datetime(2, 2, 2, 2), datetime(2, 2, 2, 2), jt.MICROSECOND, []),
        (datetime(2, 2, 2, 2), datetime(2, 2, 2, 2), -jt.MICROSECOND, [])
    ])
    def test_iter_datetime(self, start, stop, step, results):
        assert list(jt.timerange(start, stop, step)) == results

    @pytest.mark.parametrize('timerange_object', timerange_samples)
    def test_bool(self, timerange_object):
        assert bool(timerange_object) == bool(list(timerange_object))

    @pytest.mark.parametrize('timerange_object', timerange_samples)
    def test_timerange_length(self, timerange_object):
        assert len(timerange_object) == len(list(timerange_object))

    @pytest.mark.parametrize('timerange_object', timerange_samples)
    def test_to_str_and_repr(self, timerange_object):
        answer = "timerange(start={}, stop={}, step={})".format(
            repr(timerange_object.start),
            repr(timerange_object.stop),
            repr(timerange_object.step),
        )
        assert str(timerange_object) == repr(timerange_object) == answer

    def test_equal(self):
        assert jt.timerange(
            date(1, 1, 1), date(1, 1, 1)
        ) == jt.timerange(
            date(1, 1, 1), date(1, 1, 1)
        )

    def test_not_equal(self):
        assert jt.timerange(
            date(1, 1, 1), date(1, 1, 1)
        ) != jt.timerange(
            date(1, 1, 2), date(1, 1, 1)
        )

    def test_start_not_equal(self):
        assert jt.timerange(
            date(1, 1, 1), date(1, 1, 1)
        ) != jt.timerange(
            date(1, 1, 2), date(1, 1, 1)
        )

    def test_stop_not_equal(self):
        assert jt.timerange(
            date(1, 1, 2), date(1, 1, 1)
        ) != jt.timerange(
            date(1, 1, 2), date(1, 1, 2)
        )

    def test_step_not_equal(self):
        assert jt.timerange(
            date(1, 1, 2), date(1, 1, 2), jt.DAY
        ) != jt.timerange(
            date(1, 1, 2), date(1, 1, 2), jt.MINUTE
        )

    def test_type_not_equal(self):
        assert jt.timerange(
            datetime(1, 1, 1), datetime(1, 1, 1)
        ) != timedelta(1)

    def test_hash(self):
        assert hash(jt.timerange(
            date(1, 1, 1), date(1, 1, 1)
        )) == hash(jt.timerange(
            date(1, 1, 1), date(1, 1, 1)
        ))


@pytest.mark.parametrize('arg, res', [
    (datetime(2017, 11, 24, 1), datetime(2017, 11, 1, 1)),
    (datetime(1970, 1, 1, 1, 1, 1), datetime(1970, 1, 1, 1, 1, 1)),
    (date(2222, 2, 28), date(2222, 2, 1)),
])
def test_this_month_first_day(arg, res):
    assert jt.this_month_first_day(arg) == res


@pytest.mark.parametrize('arg, res', [
    (datetime(2017, 11, 24, 1), datetime(2017, 11, 30, 1)),
    (datetime(1970, 1, 1, 1, 1, 1), datetime(1970, 1, 31, 1, 1, 1)),
    (date(2222, 2, 28), date(2222, 2, 28)),
    (date(2000, 2, 23), date(2000, 2, 29)),
    (datetime(2017, 8, 3), datetime(2017, 8, 31)),
])
def test_this_month_last_day(arg, res):
    assert jt.this_month_last_day(arg) == res


@pytest.mark.parametrize('arg, res', [
    (datetime(2017, 12, 31), datetime(2017, 11, 1)),
    (datetime(2016, 1, 29, second=23), datetime(2015, 12, 1, second=23)),
    (date(2000, 1, 1), date(1999, 12, 1)),
])
def test_prev_month_first_day(arg, res):
    assert jt.prev_month_first_day(arg) == res


@pytest.mark.parametrize('arg, res', [
    (datetime(2017, 3, 31), datetime(2017, 2, 28)),
    (datetime(2016, 3, 31), datetime(2016, 2, 29)),
    (datetime(2017, 12, 31), datetime(2017, 11, 30)),
    (datetime(2016, 1, 29, second=23), datetime(2015, 12, 31, second=23)),
    (date(2000, 1, 1), date(1999, 12, 31)),
])
def test_prev_month_last_day(arg, res):
    assert jt.prev_month_last_day(arg) == res


@pytest.mark.parametrize('arg, res', [
    (datetime(2017, 1, 31), datetime(2017, 2, 1)),
    (datetime(2016, 1, 15), datetime(2016, 2, 1)),
    (datetime(2017, 12, 31), datetime(2018, 1, 1)),
    (datetime(2016, 7, 29, hour=23), datetime(2016, 8, 1, hour=23)),
    (date(2000, 1, 1), date(2000, 2, 1)),
])
def test_next_month_first_day(arg, res):
    assert jt.next_month_first_day(arg) == res


@pytest.mark.parametrize('arg, res', [
    (datetime(2017, 1, 31), datetime(2017, 2, 28)),
    (datetime(1899, 12, 15), datetime(1900, 1, 31)),
    (datetime(2017, 10, 31), datetime(2017, 11, 30)),
    (datetime(2016, 5, 9, microsecond=1), datetime(2016, 6, 30, microsecond=1)),
    (date(2000, 1, 1), date(2000, 2, 29)),
])
def test_next_month_last_day(arg, res):
    assert jt.next_month_last_day(arg) == res


tests/test_containers.py
# coding: utf-8

import pytest

from jzqt.containers import ObjectDict, FrozenDict


class TestObjectDict(object):

    def test_init(self):
        assert ObjectDict() == dict()
        assert ObjectDict({'a': 'b'}) == {'a': 'b'}
        assert ObjectDict(a=1, b=2) == dict(a=1, b=2)
        iterable_object = [('a', 1), (2, None), (None, ''), ('', False)]
        assert ObjectDict(iterable_object) == dict(iterable_object)

    def test_getattr(self):
        obj = ObjectDict({
            'name': 'JZQT',
            'get': 'data'
        })
        assert obj.name == 'JZQT'
        assert obj.get != 'data'
        assert obj['get'] == 'data'

    def test_getattr_raise_attribute_error(self):
        obj = ObjectDict()
        with pytest.raises(AttributeError) as exc_info:
            print(obj.not_exist_attr)
        assert exc_info.value.args == ('not_exist_attr',)

    def test_setattr(self):
        obj = ObjectDict()
        obj.name = 'JZQT'
        obj.get = 1
        assert obj['name'] == 'JZQT'
        assert obj['get'] == 1

    def test_delattr(self):
        obj = ObjectDict({'name': 'jzqt'})
        del obj.name
        assert obj == ObjectDict()

    def test_delattr_raise_attribute_error(self):
        obj = ObjectDict()
        with pytest.raises(AttributeError) as exc_info:
            del obj.name
        assert exc_info.value.args == ('name',)


class TestFrozenDict(object):

    _frozendict_samples = [
        FrozenDict(),
        FrozenDict({'key': 'value'}),
        FrozenDict({'key': False}),
        FrozenDict({'key': None}),
        FrozenDict({'key': True}),
        FrozenDict({'key': FrozenDict({'foo': 'bar', 'key': 1})}),
    ]

    @pytest.mark.parametrize('frozendict', _frozendict_samples)
    def test_str(self, frozendict):
        assert str(frozendict) == 'FrozenDict({})'.format(dict(frozendict))

    @pytest.mark.parametrize('frozendict', _frozendict_samples)
    def test_repr(self, frozendict):
        assert repr(frozendict) == 'FrozenDict({})'.format(dict(frozendict))

    def test_setitem(self):
        with pytest.raises(NotImplementedError) as exc_info:
            FrozenDict()['key'] = 'value'
        assert exc_info.value.args == ('FrozenDict.__setitem__',)

        with pytest.raises(NotImplementedError) as exc_info:
            FrozenDict().__setitem__('key', 'value')
        assert exc_info.value.args == ('FrozenDict.__setitem__',)

    def test_delitem(self):
        with pytest.raises(NotImplementedError) as exc_info:
            del FrozenDict()['key']
        assert exc_info.value.args == ('FrozenDict.__delitem__',)

        with pytest.raises(NotImplementedError) as exc_info:
            FrozenDict().__delitem__('key')
        assert exc_info.value.args == ('FrozenDict.__delitem__',)

    def test_clear(self):
        with pytest.raises(NotImplementedError) as exc_info:
            FrozenDict().clear()
        assert exc_info.value.args == ('FrozenDict.clear',)

    def test_update(self):
        with pytest.raises(NotImplementedError) as exc_info:
            FrozenDict().update({})
        assert exc_info.value.args == ('FrozenDict.update',)

    def test_pop(self):
        with pytest.raises(NotImplementedError) as exc_info:
            FrozenDict().pop('key')
        assert exc_info.value.args == ('FrozenDict.pop',)

    def test_popitem(self):
        with pytest.raises(NotImplementedError) as exc_info:
            FrozenDict().popitem()
        assert exc_info.value.args == ('FrozenDict.popitem',)

    def test_setdefault(self):
        with pytest.raises(NotImplementedError) as exc_info:
            FrozenDict().setdefault('key', 'value')
        assert exc_info.value.args == ('FrozenDict.setdefault',)

    def test_copy(self):
        x = FrozenDict({'a': {}})
        assert isinstance(x.copy(), FrozenDict)
        assert x['a'] is x.copy()['a']

    def test_fromkeys(self):
        res = FrozenDict.fromkeys(['a'])
        assert isinstance(res, FrozenDict)
        assert res == FrozenDict({'a': None})


tests/test_importools.py
# coding: utf-8

import pytest

from jzqt.importools import import_object


@pytest.mark.parametrize('name, obj', [
    ('pytest', pytest),
    ('pytest.raises', pytest.raises),
    ('pytest.collect', pytest.collect),
    ('jzqt.importools.import_object', import_object),
])
def test_import_object(name, obj):
    assert import_object(name) is obj


@pytest.mark.parametrize('name, msg', [
    ('not_exist_module', None),
    ('time.not_exist_object', 'cannot from time import not_exist_object'),
    ('os.not_exist_module.not_exist_object', None),
])
def test_import_object_raise_error(name, msg):
    with pytest.raises(ImportError) as exc_info:
        import_object(name)
    if msg is not None:
        assert exc_info.value.args == (msg,)


jzqt/containers.py
# coding: utf-8
"""Container data type"""


class ObjectDict(dict):
    """Dict subclass for support attribute style access.

    >>> from jzqt.containers import ObjectDict
    >>> obj = ObjectDict({'name': 'JZQT'})
    >>> obj.name
    'JZQT'
    >>> obj['name']
    'JZQT'
    >>> obj.get('name')
    'JZQT'
    """

    def __getattr__(self, name):
        try:
            return self[name]
        except KeyError:
            raise AttributeError(name)

    def __setattr__(self, name, value):
        self[name] = value


class FrozenDict(dict):
    """Immutable dict."""

    def __setitem__(self, key, value):
        """Not implemented."""
        raise NotImplementedError('FrozenDict.__setitem__')

    def __delitem__(self, key):
        """Not implemented."""
        raise NotImplementedError('FrozenDict.__delitem__')

    def __str__(self):
        return self.__repr__()

    def __repr__(self):
        return ''.join(('FrozenDict(', super(FrozenDict, self).__repr__(), ')'))

    def clear(self):
        """Not implemented."""
        raise NotImplementedError('FrozenDict.clear')

    def copy(self):
        return FrozenDict(super(FrozenDict, self).copy())

    @staticmethod
    def fromkeys(iterable, value=None):
        """Returns a new FrozenDict with keys from iterable and values equal to value."""
        return FrozenDict(dict.fromkeys(iterable, value))

    def pop(self, k, d=None):
        """Not implemented."""
        raise NotImplementedError('FrozenDict.pop')

    def popitem(self):
        """Not implemented."""
        raise NotImplementedError('FrozenDict.popitem')

    def setdefault(self, k, d=None):
        """Not implemented."""
        raise NotImplementedError('FrozenDict.setdefault')

    def update(self, *args, **kwargs):
        """Not implemented."""
        raise NotImplementedError('FrozenDict.update')


jzqt/timetools.py
# coding: utf-8
"""Date and time tools"""

from calendar import isleap
from datetime import date, datetime, timedelta

DAY = timedelta(1)
HOUR = timedelta(hours=1)
WEEK = timedelta(weeks=1)
SECOND = timedelta(seconds=1)
MINUTE = timedelta(minutes=1)
MICROSECOND = timedelta(microseconds=1)
MILLISECOND = timedelta(milliseconds=1)
ZERO_TIMEDELTA = timedelta()


class timerange(object):
    """Time range object, like ``range``

    timerange(start, stop[, step]) -> timerange object

    >>> from datetime import date, datetime, timedelta
    >>> list(timerange(datetime(1970, 1, 1), datetime(19701, 1, 3)))
    [datetime.datetime(1970, 1, 1, 0, 0), datetime.datetime(1970, 1, 2, 0, 0)]
    >>> list(timerange(datetime(1970, 1, 1, 1),
    ...      datetime(1970, 1, 1, 1, 4),
    ...      timedelta(hours=2)))
    [datetime.datetime(1970, 1, 1, 1, 0), datetime.datetime(1970, 1, 1, 3, 0)]
    >>> list(timerange(date(2017, 12, 31), date(2018, 1, 2)))
    [datetime.date(2017, 12, 31), datetime.date(2018, 1, 1)]
    """

    def __init__(self, start, stop, step=None):
        for idx, arg in ((1, start), (2, stop)):
            if not isinstance(arg, (date, datetime)):
                raise TypeError('timerange() arg {} must be date '
                                'or datetime object'.format(idx))
        if step is None:
            step = timedelta(1)
        if not isinstance(step, timedelta):
            raise TypeError('timerange() arg 3 must be timedelta object')
        if step == ZERO_TIMEDELTA:
            raise ValueError('timerange() arg 3 must not be timedelta(0)')
        self._start = start
        self._stop = stop
        self._step = step
        # compute length
        if step > ZERO_TIMEDELTA and start > stop:
            self._length = 0
        elif step < ZERO_TIMEDELTA and start < stop:
            self._length = 0
        else:
            delta_length = int((self._stop - self._start).total_seconds() * 1000000)
            step_length = int(self._step.total_seconds() * 1000000)
            self._length = int(delta_length / step_length)
            if delta_length % step_length != 0:
                self._length += 1

    def __str__(self):
        return "timerange(start={}, stop={}, step={})".format(
            repr(self._start), repr(self._stop), repr(self._step)
        )

    def __repr__(self):
        return self.__str__()

    def __bool__(self):
        return bool(self._length)

    def __len__(self):
        return self._length

    def __iter__(self):
        curr = self._start
        step = self._step
        stop = self._stop

        def should_iterate():
            if step > timedelta():
                return curr < stop
            return curr > stop

        while should_iterate():
            yield curr
            try:
                curr += step
            except OverflowError:
                break

    def __eq__(self, other):
        if not isinstance(other, timerange):
            return False
        if self._start != other.start:
            return False
        if self._stop != other.stop:
            return False
        if self._step != other.step:
            return False
        return True

    def __ne__(self, other):
        return not self.__eq__(other)

    def __hash__(self):
        return (self._start, self._stop, self._step).__hash__()

    @property
    def start(self):
        return self._start

    @property
    def stop(self):
        return self._stop

    @property
    def step(self):
        return self._step


def prev_month_first_day(dt=None):
    """Return last month first day.

    >>> from datetime import datetime, date
    >>> print(prev_month_first_day(datetime(1970, 1, 1, 11, 11)))
    1969-12-01 11:11:00
    >>> print(prev_month_first_day(date(1970, 8, 9)))
    1970-07-01
    """
    dt = dt or datetime.now()
    if dt.month == 1:
        return dt.replace(year=dt.year - 1, month=12, day=1)
    return dt.replace(month=dt.month - 1, day=1)


def prev_month_last_day(dt=None):
    """Return last month last day.

    >>> from datetime import datetime, date
    >>> print(prev_month_last_day(datetime(2017, 1, 31, 1, 1)))
    2016-12-31 01:01:00
    >>> print(prev_month_last_day(date(1900, 3, 23)))
    1900-02-29
    """
    dt = dt or datetime.now()
    if dt.month == 1:
        return this_month_last_day(dt.replace(year=dt.year - 1, month=12))
    return this_month_last_day(dt.replace(month=dt.month - 1, day=1))


def this_month_first_day(dt=None):
    """Return this month first day.

    >>> from datetime import datetime, date
    >>> print(this_month_first_day(datetime(2222, 12, 31, 11, 22, 33)))
    2222-12-01 11:22:33
    >>> print(this_month_first_day(date(1970, 1, 1)))
    1970-01-01
    """
    dt = dt or datetime.now()
    return dt.replace(day=1)


def this_month_last_day(dt=None):
    """Return this month last day.

    >>> from datetime import datetime, date
    >>> print(this_month_last_day(datetime(2000, 2, 9, 3)))
    2000-02-29 03:00:00
    >>> print(this_month_last_day(date(1999, 8, 3)))
    1999-08-31
    """
    dt = dt or datetime.now()
    if dt.month in {1, 3, 5, 7, 8, 10, 12}:
        return dt.replace(day=31)
    if dt.month != 2:
        return dt.replace(day=30)
    if isleap(dt.year):
        return dt.replace(day=29)
    return dt.replace(day=28)


def next_month_first_day(dt=None):
    """Return next month first day.

    >>> from datetime import datetime, date
    >>> print(next_month_first_day(datetime(1999, 12, 9, 3)))
    2000-01-01 03:00:00
    >>> print(next_month_first_day(date(1999, 7, 3)))
    1999-08-01
    """
    dt = dt or datetime.now()
    if dt.month == 12:
        return dt.replace(year=dt.year + 1, month=1, day=1)
    return dt.replace(month=dt.month + 1, day=1)


def next_month_last_day(dt=None):
    """Return next month last day.

    >>> from datetime import datetime, date
    >>> print(next_month_last_day(datetime(1999, 12, 9, 3)))
    2000-01-31 03:00:00
    >>> print(next_month_last_day(date(1999, 1, 3)))
    1999-02-28
    """
    dt = dt or datetime.now()
    if dt.month == 12:
        return this_month_last_day(dt.replace(year=dt.year + 1, month=1))
    return this_month_last_day(dt.replace(month=dt.month + 1, day=1))
